core/src/main/scala/org/apache/spark/RecoverCacheShutdown.scala

```diff
@@ -0,0 +1,234 @@
+import org.apache.spark.util.ThreadUtils
+
+/**
+ * Responsible for asynchronously replicating all of an executors cached blocks, and then shutting
+ * it down.
+ */
+final private class RecoverCacheShutdown(
+    state: RecoverCacheShutdownState,
+    conf: SparkConf)
+  extends Logging {
+
+  private val threadPool = ThreadUtils.newDaemonCachedThreadPool("recover-cache-shutdown-pool")
+  private implicit val asyncExecutionContext = ExecutionContext.fromExecutorService(threadPool)
+
+  /**
+   * Start the recover cache shutdown process for these executors
+   *
+   * @param execIds the executors to start shutting down
+   */
+  def startExecutorKill(execIds: Seq[String]): Unit = {
+    logDebug(s"Recover cached data before shutting down executors ${execIds.mkString(", ")}.")
+    checkForReplicableBlocks(execIds)
+  }
+
+  /**
+   * Stops all thread pools
+   *
+   * @return
+   */
+  def stop(): java.util.List[Runnable] = {
+    threadPool.shutdownNow()
+    state.stop()
+  }
+
+  /**
+   * Get list of cached blocks from BlockManagerMaster. If there are cached blocks, replicate them,
+   * otherwise kill the executors
+   *
+   * @param execIds the executors to check
+   */
+  private def checkForReplicableBlocks(execIds: Seq[String]) = state.getBlocks(execIds).foreach {
+    case (executorId, NoMoreBlocks) => state.killExecutor(executorId)
+    case (executorId, NotEnoughMemory) => state.killExecutor(executorId)
+    case (executorId, HasCachedBlocks) => replicateBlocks(executorId)
+  }
+
+  /**
+   * Replicate one cached block on an executor. If there are more, repeat. If there are none, check
+   * with the block manager master again. If there is an error, go ahead and kill executor.
+   *
+   * @param execId the executor to save a block one
+   */
+  private def replicateBlocks(execId: String): Unit = {
+    val (response, blockId) = state.replicateFirstBlock(execId)
+    response.onComplete {
+      case scala.util.Success(true) =>
+        logTrace(s"Finished replicating block ${blockId.getOrElse("unknown")} on exec $execId.")
+        replicateBlocks(execId)
+      case scala.util.Success(false) => checkForReplicableBlocks(Seq(execId))
+      case Failure(f) =>
+        logWarning(s"Error trying to replicate block ${blockId.getOrElse("unknown")}.", f)
+        state.killExecutor(execId)
+    }
+  }
+}
+
+private object RecoverCacheShutdown {
+  def apply(eam: ExecutorAllocationManager, conf: SparkConf): RecoverCacheShutdown = {
+    val bmme = SparkEnv.get.blockManager.master.driverEndpoint
+    val state = new RecoverCacheShutdownState(bmme, eam, conf)
+    new RecoverCacheShutdown(state, conf)
+  }
+}
+
+/**
+ * Private class that holds state for all the executors being shutdown.
+ * @param blockManagerMasterEndpoint blockManagerMasterEndpoint
+ * @param executorAllocationManager ExecutorAllocationManager
+ * @param conf spark conf
+ */
+final private class RecoverCacheShutdownState(
```

Comment:
Does this need to be a separate class? Seems like the code could easily live inside `RecoverCacheShutdown`.

Reply:
These could be the same class. I thought keeping the state mutating methods together in their own class would be a little easier to understand.